Casm 2787 add remove ncns 1.2 (#1536)
Cray-HPE/docs-csm: operations/utility_storage/Add_Ceph_Node.md

    ```

    **IMPORTANT:** In the output from `watch ceph -s` the health should go to a `HEALTH_WARN` state. This is expected. Most commonly you will see an alert about `failed to probe daemons or devices`, but this should clear on its own.

>     In addition, it may take up to 5 minutes for the added OSDs to report as `up`.  This is dependent on the Ceph Orchestrator performing an inventory and completing batch processing to add the OSDs.

Reviewer: ```suggestion
    In addition, it may take up to 5 minutes for the added OSDs to report as `up`. This is dependent on the Ceph Orchestrator performing an inventory and completing batch processing to add the OSDs.
```

## Regenerate Rados-GW Load Balancer Configuration for the Rebuilt Nodes

>    **IMPORTANT:** `radosgw` by default is deployed to the first 3 storage nodes. This includes `haproxy` and `keepalived`.

Reviewer: ```suggestion
   **IMPORTANT:** `radosgw` by default is deployed to the first three storage nodes. This includes `haproxy` and `keepalived`.
```

>    This is automated as part of the install, but you may have to regenerate the configuration if you are not running on the first 3 storage nodes or all nodes.

Reviewer: ```suggestion
   This is automated as part of the install, but the configuration may need to be regenerated if not running on the first three storage nodes or all nodes.
```

1. Deploy Rados Gateway containers to the new nodes.

>    - If running Rados Gateway on all nodes is the desired configuration then run:

Reviewer: ```suggestion
   - If running Rados Gateway on all nodes is the desired configuration, then run:
```

      ```bash
      ncn-s00(1/2/3)# ceph orch apply rgw site1 zone1 --placement="*" --port=8080
      ```

>    - If deploying to select nodes then do:

Reviewer: ```suggestion
   - If deploying to select nodes, then instead run:
```

                systemctl restart keepalived.service'
     ```

   - If the node was added:

     Determine the IP address of the added node.

     ```bash
     ncn-s# cloud-init query ds | jq -r ".meta_data[].host_records[] | select(.aliases[]? == \"$(hostname)\") | .ip" 2>/dev/null
     ```

     Example Output:

     ```text
     10.252.1.13
     ```

>      Update the HAproxy config to include the added node. Select a storage node `ncn-s00x` from `ncn-s001`, `ncn-s002`, or `ncn-s003`. This cannot be done from the added node.

Reviewer: ```suggestion
     Update the HAproxy configuration to include the added node. Select one of the first three storage nodes. This cannot be done from the added node.
```

     ```bash

>      ncn-s00x# vi /etc/haproxy/haproxy.cfg

Reviewer: ```suggestion
     ncn-s# vi /etc/haproxy/haproxy.cfg
```

     ```

     This example adds or updates `ncn-s004` with the IP address `10.252.1.13` to `backend rgw-backend`.

     ```text
     ...
     backend rgw-backend
         option forwardfor
         balance static-rr
         option httpchk GET /
             server server-ncn-s001-rgw0 10.252.1.6:8080 check weight 100
             server server-ncn-s002-rgw0 10.252.1.5:8080 check weight 100
             server server-ncn-s003-rgw0 10.252.1.4:8080 check weight 100
             server server-ncn-s004-rgw0 10.252.1.13:8080 check weight 100   <--- Added or updated line
     ...
     ```

>      Copy the updated HAproxy config to all the storage nodes. Adjust the command based on the number of storage nodes.

Reviewer: ```suggestion
     Copy the updated HAproxy configuration to all the storage nodes. Adjust the command based on the number of storage nodes.
```

     ```bash
     ncn-s00x# pdcp -w ncn-s00[1-(end node number)] /etc/haproxy/haproxy.cfg /etc/haproxy/haproxy.cfg
     ```

     Configure `apparmor` and KeepAlived **on the added node** and restart the services across all the storage nodes.

     ```bash
     ncn-s# source /srv/cray/scripts/metal/update_apparmor.sh; reconfigure-apparmor
     ncn-s# /srv/cray/scripts/metal/generate_keepalived_conf.sh > /etc/keepalived/keepalived.conf
     ncn-s# export  PDSH_SSH_ARGS_APPEND="-o StrictHostKeyChecking=no"
     ncn-s# pdsh -w ncn-s00[1-(end node number)] -f 2 'systemctl restart haproxy.service; systemctl restart keepalived.service'
     ```

## Next Step

> - If rebuilding the storage node, proceed to [Storage Node Validation](../node_management/Rebuild_NCNs/Post_Rebuild_Storage_Node_Validation.md)

Reviewer: ```suggestion
- If rebuilding the storage node, then proceed to [Storage Node Validation](../node_management/Rebuild_NCNs/Post_Rebuild_Storage_Node_Validation.md).
```

> - If adding the storage node, return to [Boot NCN - Add storage node to the Ceph cluster](../node_management/Add_Remove_Replace_NCNs/Boot_NCN.md#boot-ncn-storage-nodes-only) for the next step

Reviewer: ```suggestion
- If adding the storage node, then return to [Boot NCN - Add storage node to the Ceph cluster](../node_management/Add_Remove_Replace_NCNs/Boot_NCN.md#boot-ncn-storage-nodes-only) for the next step.
```
